### PR TITLE
feat: implement query API for RegionServer

### PR DESCRIPTION
**Pull request by waynexia**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Changes:

- Define `RegionServer::handle_read()`
- Define dummy structs for DataFusion
- Move `TableFactory` and `DataSource` to store-api

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

**Review by fengjiachun — APPROVED** [review]

File: src/datanode/src/region_server.rs [hunk 1]
```diff
@@ -112,13 +161,150 @@ enum RegionChange {
...
 impl DummyCatalogList {
-    pub fn new(region_id: RegionId) -> Self {
-        todo!()
+    pub async fn new(region_id: RegionId, engine: RegionEngineRef) -> Result<Self> {
+        let metadata =
+            engine
+                .get_metadata(region_id)
+                .await
+                .with_context(|_| GetRegionMetadataSnafu {
+                    region: engine.name(),
```
Comment (evenyag) [inline src/datanode/src/region_server.rs:177]: Isn't it `engine`?
```suggestion
                    engine: engine.name(),
```

File: src/datanode/src/region_server.rs [hunk 2]
```diff
@@ -112,13 +161,150 @@ enum RegionChange {
...
+    }
+
+    async fn scan(
+        &self,
+        _state: &SessionState,
+        projection: Option<&Vec<usize>>,
+        filters: &[DfExpr],
+        limit: Option<usize>,
+    ) -> DfResult<Arc<dyn DfPhysicalPlan>> {
+        let mut request = self.scan_request.lock().unwrap().clone();
```
Comment (evenyag) [inline src/datanode/src/region_server.rs:296]: question: Why do we put a scan request in the provider? Could we create a fresh scan request each time?

File: src/datanode/src/region_server.rs [hunk 3]
```diff
@@ -112,13 +161,150 @@ enum RegionChange {
     Deregisters,
 }
 
-#[allow(dead_code)]
-struct DummyCatalogList {}
+/// Resolve to the given region (specified by [RegionId]) unconditionally.
+#[derive(Clone)]
+struct DummyCatalogList {
```
Comment (evenyag) [inline src/datanode/src/region_server.rs:166]: Consider putting them in a `dummy` mod.

File: src/store-api/src/region_engine.rs [hunk 4]
```diff
@@ -44,11 +43,11 @@ pub trait RegionEngine {
     async fn handle_query(
         &self,
         region_id: RegionId,
-        request: QueryRequest,
+        request: ScanRequest,
     ) -> Result<SendableRecordBatchStream, BoxedError>;
 
     /// Retrieve region's metadata.
-    async fn get_metadata(&self, region_id: RegionId) -> Result<RegionMetadata, BoxedError>;
+    async fn get_metadata(&self, region_id: RegionId) -> Result<RegionMetadataRef, BoxedError>;
```
Comment (evenyag) [inline src/store-api/src/region_engine.rs:50]: What if the region doesn't exist? Will this method return an error?

File: src/datanode/src/region_server.rs (hunk 2 above)
Comment (waynexia) [reply src/datanode/src/region_server.rs:296]: Keeping a request makes it possible to change the request field before calling `scan`

File: src/datanode/src/region_server.rs (hunk 3 above)
Comment (waynexia) [reply src/datanode/src/region_server.rs:166]: Maybe do it later

File: src/store-api/src/region_engine.rs (hunk 4 above)
Comment (waynexia) [reply src/store-api/src/region_engine.rs:50]: Yes, the implement can return an error if it doesn't exist

**Review by evenyag — APPROVED** [review]

**Pull request merged**